impl/src/main/java/slack/ReactionMessageHandler.java

```diff
@@ -3,81 +3,125 @@
     private static final String REACTION_REMOVED = "reaction_removed";
+    static final         String MESSAGE          = "message";
+    static final         String ITEM             = "item";
+    static final         String TYPE             = "type";
+    static final         String REACTION         = "reaction";
+    static final         String CHANNEL          = "channel";
+    static final         String THREAD           = "ts";
 
-    private static final ImmutableList    HANDLED_TYPES   = ImmutableList.of(REACTION_ADDED, REACTION_REMOVED);
-    private static final ImmutableList    POSITIVE_EMOJIS = ImmutableList.of("+1");
-    private static final ImmutableList    NEGATIVE_EMOJIS = ImmutableList.of("-1");
-    private final        QuestionResource questionResource;
-    private final        AnswerResource   answerResource;
+    private static final ImmutableList      HANDLED_TYPES   = ImmutableList.of(REACTION_ADDED, REACTION_REMOVED);
+    private static final ImmutableList      POSITIVE_EMOJIS = ImmutableList.of("+1");
+    private static final ImmutableList      NEGATIVE_EMOJIS = ImmutableList.of("-1");
+    private final        QuestionResource   questionResource;
+    private final        AnswerResource     answerResource;
+    private final        UserAnswerResource userAnswerResource;
+    private final        SlackResource      slackResource;
 
     @Inject
-    public ReactionMessageHandler(QuestionResource questionResource, AnswerResource answerResource) {
+    public ReactionMessageHandler(SlackResource slackResource,
+        QuestionResource questionResource,
+        AnswerResource answerResource,
+        UserAnswerResource userAnswerResource
+    ) {
 
+        this.slackResource = slackResource;
         this.questionResource = questionResource;
         this.answerResource = answerResource;
+        this.userAnswerResource = userAnswerResource;
     }
 
     @Override
     public boolean shouldHandle(String type, JsonObject body) {
-        return HANDLED_TYPES.contains(type);
+        return HANDLED_TYPES.contains(type) && MESSAGE.equals(body.get(ITEM).getAsJsonObject().get(TYPE).getAsString());
     }
 
     @Override
     public Observable<Void> handleMessage(JsonObject message) {
-        final String threadId = message.get("item").getAsJsonObject().get("ts").getAsString();
 
-        final AtomicReference<Boolean> upVote = new AtomicReference<>();
-        String reaction = message.get("reaction").getAsString();
+        final Boolean upVote = getUpVoteValue(message);
 
-        if(POSITIVE_EMOJIS.contains(reaction)) {
-            upVote.set(true);
+        //No recognized reaction
+        if (upVote == null) {
+            return empty();
         }
 
-        if(NEGATIVE_EMOJIS.contains(reaction)) {
-            upVote.set(false);
+        final String threadId = getThread(message);
+        return slackResource.getMessageFromSlack(getChannel(message), getThread(message))
```

Comment:
The nesting is kind of hard to read. Perhaps refactor by extracting the flatmap Func1's to variables?

Reply:
Yup. That will make it a bit better.